applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_face_load_condition.cpp

```diff
@@ -72,134 +72,65 @@ void UPwNormalFaceLoadCondition<TDim,TNumNodes>::
+template<unsigned int TDim, unsigned int TNumNodes>
+void UPwNormalFaceLoadCondition<TDim, TNumNodes>::InitializeConditionVariables(
+    NormalFaceLoadVariables& rVariables,
+    const GeometryType& rGeom)
 {
-    for (unsigned int i=0; i<2; ++i) {
-        rVariables.NormalStressVector[i]     = rGeom[i].FastGetSolutionStepValue(NORMAL_CONTACT_STRESS);
-        rVariables.TangentialStressVector[i] = rGeom[i].FastGetSolutionStepValue(TANGENTIAL_CONTACT_STRESS);
-    }
-}
-
-//----------------------------------------------------------------------------------------
-template< >
-void UPwNormalFaceLoadCondition<3,3>::
-    InitializeConditionVariables(NormalFaceLoadVariables& rVariables,
-                                 const GeometryType& rGeom)
-{
-    for (unsigned int i=0; i<3; ++i) {
+    for (unsigned int i = 0; i < TNumNodes; ++i) {
         rVariables.NormalStressVector[i] = rGeom[i].FastGetSolutionStepValue(NORMAL_CONTACT_STRESS);
     }
-}
-
-//----------------------------------------------------------------------------------------
-template< >
-void UPwNormalFaceLoadCondition<3,4>::
-    InitializeConditionVariables(NormalFaceLoadVariables& rVariables,
-                                 const GeometryType& rGeom)
-{
-    for (unsigned int i=0; i<4; ++i) {
-        rVariables.NormalStressVector[i] = rGeom[i].FastGetSolutionStepValue(NORMAL_CONTACT_STRESS);
+    //
+    if (TDim == 2) {
+        for (unsigned int i = 0; i < TNumNodes; ++i) {
+            rVariables.TangentialStressVector[i] = rGeom[i].FastGetSolutionStepValue(TANGENTIAL_CONTACT_STRESS);
+        }
     }
 }
 
-//----------------------------------------------------------------------------------------
-template< >
-void UPwNormalFaceLoadCondition<2,2>::
-    CalculateTractionVector(array_1d<double,2>& rTractionVector,
-                            const Matrix& Jacobian,
-                            const Matrix& NContainer,
-                            const NormalFaceLoadVariables& Variables,
-                            const unsigned int& GPoint)
+// ============================================================================================
+// ============================================================================================
+template<unsigned int TDim, unsigned int TNumNodes>
+void UPwNormalFaceLoadCondition<TDim, TNumNodes>::CalculateTractionVector(
+    array_1d<double, TDim>& rTractionVector,
+    const Matrix& Jacobian,
+    const Matrix& NContainer,
+    const NormalFaceLoadVariables& Variables,
+    const unsigned int& GPoint)
 {
-    double NormalStress = 0.0;
-    double TangentialStress = 0.0;
-
-    for (unsigned int i=0; i<2; ++i) {
-        NormalStress     += NContainer(GPoint,i)*Variables.NormalStressVector[i];
-        TangentialStress += NContainer(GPoint,i)*Variables.TangentialStressVector[i];
+    Vector NormalVector = ZeroVector(TDim);
+    double NormalStress = MathUtils<>::Dot(row(NContainer, GPoint), Variables.NormalStressVector);
```

Comment:
Due to your improvement, this can be const  (very nice improvement btw, way better than the old loop)
```suggestion
    const double NormalStress = MathUtils<>::Dot(row(NContainer, GPoint), Variables.NormalStressVector);
```